### DISCUSSION TITLE
iCub landed on ROS2 planet

### DISCUSSION THREAD

**Discussion by Nicogene** [robotology/community#610] [Research news]
Hi @robotology/everyone!

Thanks to a joint work between [HSP@IIT](https://www.iit.it/web/humanoid-sensing-and-perception) and [iCub-Tech](https://www.iit.it/web/icub-tech) department iCub humanoid robot is now ROS2 compatible.

These type of data can be now published on topic ros2 using the devices contained in the [yarp-devices-ros2](https://github.com/robotology-playground/yarp-devices-ros2):
- [Joint state](http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/JointState.html) 
- [IMU](http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/Imu.html)
- [FT sensors](http://docs.ros.org/en/noetic/api/geometry_msgs/html/msg/Wrench.html)
- [Laser scan](http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/LaserScan.html)
- [Image](http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/Image.html)
- [Occupancy Grid](http://docs.ros.org/en/noetic/api/nav_msgs/html/msg/OccupancyGrid.html)
- [Point cloud](http://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/PointCloud2.html)
- [Odometry](http://docs.ros.org/en/noetic/api/nav_msgs/html/msg/Odometry.html)

Here is an example of a visualization of iCub in `rviz2` using the gazebo simulator

https://user-images.githubusercontent.com/19152494/213452515-dd5d387b-8059-44b0-8b59-5ac8de579a1c.mp4

Here instead an example of `rviz2` visualization of the real robot performing a grasping demo (credits @xEnVrE)

https://user-images.githubusercontent.com/19152494/216341281-a8b6abd8-3445-4b3b-ae11-5ea13473a7c9.mp4

The part on data publication on topic ros2 is almost done, we miss just the skin sensors.
 
On the other hand, the control part is still in an experimental state. We hope to give you news soon also on this matter.

> [!warning]
**`yarp-devices-ros2` is still experimental/beta, as also its README states. It will be "officially" released after YARP 3.8.0 release which will happen this month. Please take account that breaking changes could be done without notice.** 

If you have an iCub and you want to try out this ROS2 integration, you need:

- Install [yarp-devices-ros2](https://github.com/robotology-playground/yarp-devices-ros2)
- Change the `yarprobotinterface` files like done here: https://github.com/robotology/robots-configuration/pull/415/files

Special thanks to @randaz81 @elandini84 and @ste93.

**Comment by lrapetti** [comment]
> These type of data can be now published on topic ros2 using the devices contained in the [yarp-devices-ros2](https://github.com/robotology-playground/yarp-device-ros2):

There is a typo in the link https://github.com/robotology-playground/yarp-devices-ros2 😉

**Reply by Nicogene** [reply]
Thanks!

**Comment by traversaro** [comment]
> 
> If you have an iCub and you want to try out this ROS2 integration, you need:
> 
>     * Install [yarp-devices-ros2](https://github.com/robotology-playground/yarp-devices-ros2)
> 
>     * Change the `yarprobotinterface` files like done here: https://github.com/robotology/robots-configuration/pull/415/files
> 

Thanks a lot @Nicogene ! How it is possible to try this integration using the Gazebo Classic simulation of iCub, shown in the first video?

**Reply by Nicogene** [reply]
You can find [here](https://github.com/robotology/icub-models/tree/devel/iCub_manual/ros2) an (I hope exhaustive) guide about how to use the ros2 integration with the Gazebo simulator. If you find any error or missing information don't hesitate to open an issue in the [icub-models](https://github.com/robotology/icub-models) repository!

**Reply by alecive** [reply]
You are so classic @traversaro

**Comment by xEnVrE** [comment]
> You can find [here](https://github.com/robotology/icub-models/tree/devel/ros2) an (I hope exhaustive) guide about how to use the ros2 integration with the Gazebo simulator.

It seems that the link above is broken. Was it moved somewhere else? cc @Nicogene @traversaro

**Reply by martinaxgloria** [reply]
Hi @xEnVrE,
I'm not sure it's the same file, but following the history, I think you can find the guide [here](https://github.com/robotology/icub-models/tree/devel/iCub_manual/ros2)

**Reply by xEnVrE** [reply]
Thank you!

**Reply by Nicogene** [reply]
I changed it w/ the link provided by @martinaxgloria 👍🏻
